src/fetcher.ts

```diff
@@ -1,78 +1,91 @@
+    this.version = clientOptions.version || this.version
+    this.timeout = clientOptions.timeout || 5000
+    this.maxRetries = clientOptions.maxRetries || 2
   }
 
-  private getURL(path: string): string {
-    return `${this.baseURL}${path}`
+  private getApiKey(): string {
+    return this.apiKey
   }
 
-  private getAuthorizationHeader(): string {
-    return `Bearer ${this.apiKey}`
+  public getVersion(): PriceEngineVersion {
+    return this.version
   }
 
-  async get<T>(url: string, options?: RequestInit): Promise<T> {
-    const response = await this.fetcher(this.getURL(url), {
-      ...options,
-      method: 'GET',
-      headers: {
-        'Content-Type': 'application/json',
-        Authorization: this.getAuthorizationHeader(),
-        ...options?.headers
-      }
-    })
+  private getURL(endpoint: string): string {
+    return `${this.baseURL}${endpoint}`
+  }
 
-    return response.json()
+  private getAuthorizationHeader(): string {
+    return `Bearer ${this.getApiKey()}`
   }
 
-  async post<T>(url: string, body: any, options?: RequestInit): Promise<T> {
-    const response = await this.fetcher(this.getURL(url), {
+  private async requestWithTimeout<T>(
+    url: string,
+    options: RequestInit
+  ): Promise<T> {
+    const controller = new AbortController()
+    const id = setTimeout(() => controller.abort(), this.timeout)
+    const response: Response = await fetch(url, {
       ...options,
-      method: 'POST',
-      body: JSON.stringify(body),
-      headers: {
-        'Content-Type': 'application/json',
-        Authorization: this.getAuthorizationHeader(),
-        ...options?.headers
-      }
+      signal: controller.signal
     })
-
-    return response.json()
+    clearTimeout(id)
+    return response.json() as T
   }
 
-  async put<T>(url: string, body: any, options?: RequestInit): Promise<T> {
-    const response = await this.fetcher(this.getURL(url), {
+  protected async request<T>(
+    endpoint: string,
+    options?: RequestInit
+  ): Promise<T> {
+    const url: string = this.getURL(endpoint)
+    const headers = {
+      Authorization: this.getAuthorizationHeader(),
+      'Content-Type': 'application/json'
+    }
+    const config: RequestInit = {
       ...options,
-      method: 'PUT',
-      body: JSON.stringify(body),
-      headers: {
-        'Content-Type': 'application/json',
-        Authorization: this.getAuthorizationHeader(),
-        ...options?.headers
-      }
-    })
+      headers
+    }
 
-    return response.json()
+    for (let attempt = 0; attempt <= this.maxRetries; attempt++) {
+      try {
+        const response = await this.requestWithTimeout<Response>(url, config)
+        if (response.ok) {
+          return response.json() as T
+        }
+        throw new Error(response.statusText)
+      } catch (error) {
+        if (attempt === this.maxRetries) {
+          throw error
+        }
+      }
+    }
+    throw new Error('Request failed after max retries')
```

Comment:
```suggestion
    throw new ApiError('Request failed after max retries')
```

We can use the custom api error for this